src/reflect/scala/reflect/internal/FreshNames.scala

```diff
@@ -23,17 +24,20 @@ trait FreshNames { self: Names with StdNames =>
   // Extractor that matches names which were generated by some
   // FreshNameCreator with known prefix. Extracts user-specified
   // prefix that was used as a parameter to newName by stripping
-  // global creator prefix and unique number in the end of the name.
+  // global creator prefix and unique numerical suffix.
+  // The creator prefix and numerical suffix may both be empty.
   class FreshNameExtractor(creatorPrefix: String = "") {
-    // quote prefix so that it can be used with replaceFirst
-    // which expects regExp rather than simple string
-    val quotedCreatorPrefix = java.util.regex.Pattern.quote(creatorPrefix)
-
-    def unapply(name: Name): Option[String] = {
-      val sname = name.toString
-      // name should start with creatorPrefix and end with number
-      if (!sname.startsWith(creatorPrefix) || !sname.matches("^.*\\d*$")) None
-      else Some(NameTransformer.decode(sname.replaceFirst(quotedCreatorPrefix, "").replaceAll("\\d*$", "")))
+
+    // name should start with creatorPrefix and end with number
+    val freshlyNamed = {
+      val pre = if (!creatorPrefix.isEmpty) Regex quote creatorPrefix else ""
+      s"""$pre(.*?)\\d*""".r
     }
+
+    def unapply(name: Name): Option[String] =
+      name.toString match {
+        case freshlyNamed(prefix) => Some(prefix)
```

Comment:
Friday musing: wouldn't it be nice to have a `toOption` on `List` that gives you `Some(headIHaveNoTail)` or `None`?
```
def unapply(name: Name): Option[String] = freshlyNamed.unapplySeq(name).flatMap(_.toOption)
```